text/0011-block-dissemination.md

```diff
@@ -0,0 +1,215 @@
+2. **Inadequate when peers have diverged ledger heights**: When a new peer joins the network its ledger is significantly behind the rest, yet it still receives blocks from peers in its organization which only end up discarded because it has no short term use for them.
+3. **Memory overhead**: While gossip is an ephemeral component (no way to persist data to stable storage), it often receives blocks out of order due to the random nature of the epidemic style block dissemination. Consequently, it stores blocks in an in-memory sliding window which evicts old blocks in favor of newer blocks. Even though blocks are eventually purged from memory, it is not immediate and therefore every channel the peer participates in, consumes memory throughout the runtime of the peer. This trivially limits the amount of channels a peer can be part of, and increases the memory consumption of the peer.
+4. **Complications** of diagnostics: Making only a single peer per organization to retrieve blocks makes it that network administrators and site reliability engineers need to determine which is the leader peer when troubleshooting issues related to peers lagging behind.
+5. **Non determinism**: The non deterministic and highly concurrent nature of epidemic based block replication makes it harder to build stable tests.
+
+**Point to point block replication**: In case a peer is too far behind the rest, as in situations where it is just added to the network, it may choose a peer to initiate synchronization from in a point to point manner.
+This entails a series of requests and responses where the peer behind asks for several consecutive blocks, and the peer ahead sends them back.
+Due to the idle time spent in waiting for blocks to be sent or for requests to be sent.
+
+
+This protocol is slower than one where the peer retrieves blocks from the orderer. In the latter, the orderer can just send as many blocks per second as the gRPC stream throttling allows, thus maximizing the block transfer rate.
+
+
+# The proposed new block dissemination method
+[Proposed]: #Proposed
+
+An alternative block dissemination method is proposed to replace the current block dissemination.
+
+**Note:** Whenever an "orderer" is mentioned, we mean some orderer among the available orderers via some selection policy that is orthogonal to this proposal.
+
+The proposal is to be implemented in two phases:
+
+**Phase I**: Gossip block replication among peers will cease to operate, and the relevant code will be removed. This means that both gossip block replication layer and leader election layers will be removed and each peer will always pull only from the ordering service.
```

Comment:
Question here:
Assuming we have 5 peers(from different org) and 5 orderers(running in raft).
Will those 5 peers pull block from raft leader or those 5 peers able to pull block from any or orderer nodes?

Reply:
The peers don't know who is the leader anyway, so they have to use an arbitrary node